server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java

```diff
@@ -84,31 +84,50 @@ void moveShards() {
         Queue<RoutingNode> excludedNodes = new ArrayDeque<>();
         classifyNodesForShardMovement(eligibleNodes, excludedNodes);
 
-        if (excludedNodes.isEmpty()) {
-            logger.debug("No excluded nodes found. Returning...");
-            return;
+        // move shards that cannot remain on eligible nodes
+        final List<ShardRouting> forceMoveShards = new ArrayList<>();
+        eligibleNodes.forEach(sourceNode -> {
+            for (final ShardRouting shardRouting : sourceNode) {
+                if (ineligibleForMove(shardRouting)) {
+                    continue;
+                }
+
+                if (allocation.deciders().canRemain(shardRouting, sourceNode, allocation) == Decision.NO) {
+                    forceMoveShards.add(shardRouting);
+                }
+            }
+        });
+        for (final ShardRouting shard : forceMoveShards) {
+            if (eligibleNodes.isEmpty()) {
+                logger.trace("there are no eligible nodes available, return");
+                return;
+            }
+
+            tryShardMovementToEligibleNode(eligibleNodes, shard);
         }
 
+        // move shards that are currently assigned on excluded nodes
         while (!eligibleNodes.isEmpty() && !excludedNodes.isEmpty()) {
             RoutingNode sourceNode = excludedNodes.poll();
-            for (ShardRouting ineligibleShard : sourceNode) {
-                if (ineligibleShard.started() == false) {
-                    continue;
-                }
-
-                if (!RoutingPool.REMOTE_CAPABLE.equals(RoutingPool.getShardPool(ineligibleShard, allocation))) {
+            for (final ShardRouting ineligibleShard : sourceNode) {
+                if (ineligibleForMove(ineligibleShard)) {
                     continue;
                 }
 
                 if (eligibleNodes.isEmpty()) {
-                    break;
+                    logger.trace("there are no eligible nodes available, return");
+                    return;
                 }
 
                 tryShardMovementToEligibleNode(eligibleNodes, ineligibleShard);
             }
         }
     }
 
+    private boolean ineligibleForMove(ShardRouting shard) {
+        return !shard.started() || !RoutingPool.REMOTE_CAPABLE.equals(RoutingPool.getShardPool(shard, allocation));
```

Comment:
The prevailing style is `shard.started() == false || RoutingPool.REMOTE_CAPABLE.equals(RoutingPool.getShardPool(shard, allocation) == false`